MoreLinq/AssertCount.cs

```diff
@@ -83,12 +83,21 @@ static Exception OnAssertCountFailure(int cmp, int count)
             if (count < 0) throw new ArgumentOutOfRangeException(nameof(count));
             if (errorSelector == null) throw new ArgumentNullException(nameof(errorSelector));
 
+            // Optimization for collections
             if (source is ICollection<TSource> collection)
             {
                 return collection.Count == count
                      ? collection
                      : From<TSource>(() => throw errorSelector(collection.Count.CompareTo(count), count));
-            }
+            };
+#if IREADONLY
+            if (source is IReadOnlyCollection<TSource> readOnlyCollection)
+            {
+                return readOnlyCollection.Count == count
```

Comment:
This code has non-trivial duplication with the optimisation for `ICollection<>`. The two blocks should be consolidated via a local function, as in:

```c#
        static IEnumerable<TSource> AssertCountImpl<TSource>(IEnumerable<TSource> source, 
            int count, Func<int, int, Exception> errorSelector)
        {
            if (source == null) throw new ArgumentNullException(nameof(source));
            if (count < 0) throw new ArgumentOutOfRangeException(nameof(count));
            if (errorSelector == null) throw new ArgumentNullException(nameof(errorSelector));

            switch (source)
            {
                case ICollection<TSource> collection:
                    return AssertKnownCount(collection.Count);
#if IREADONLY
                case IReadOnlyCollection<TSource> readOnlyCollection:
                    return AssertKnownCount(readOnlyCollection.Count);
#endif
                default:
                    return _(); IEnumerable<TSource> _()
                    {
                        var iterations = 0;
                        foreach (var element in source)
                        {
                            iterations++;
                            if (iterations > count)
                                throw errorSelector(1, count);
                            yield return element;
                        }
                        if (iterations != count)
                            throw errorSelector(-1, count);
                    }
            }

            IEnumerable<TSource> AssertKnownCount(int actualCount) =>
                actualCount == count
                ? source
                : From<TSource>(() => throw errorSelector(actualCount.CompareTo(count), count));
        }
```